File: Dart/CSharpLexerBase.dart
import 'package:antlr4/antlr4.dart';
import 'dart:io';
import 'dart:convert';
import 'CSharpLexer.dart';

abstract class CSharpLexerBase extends Lexer
{
    int interpolatedStringLevel;
    List<bool> interpolatedVerbatiums = [];
    List<int> curlyLevels = [];
    bool verbatium;

    CSharpLexerBase(CharStream input) : super(input)
    {
    }

    void A1()
    {
        interpolatedStringLevel++;
        interpolatedVerbatiums.add(false);
        verbatium = false;
    }

    void A2()
    {
        interpolatedStringLevel++;
        interpolatedVerbatiums.add(true);
        verbatium = true;
    }

    void A3()
    {
        if (interpolatedStringLevel > 0)
        {
            curlyLevels.add(curlyLevels.removeLast() + 1);
        }
    }

    void A4()
    {
        if (interpolatedStringLevel > 0)
        {
            curlyLevels.add(curlyLevels.removeLast() - 1);
            if (curlyLevels.last == 0)
            {
                curlyLevels.removeLast();
                skip();
                popMode();
            }
        }
    }

    void A5()
    {
        if (interpolatedStringLevel > 0)
        {
            int ind = 1;
            bool switchToFormatString = true;
            while (inputStream.LA(ind) != '}')
            {
                if (inputStream.LA(ind) == ':' || inputStream.LA(ind) == ')')
                {
                    switchToFormatString = false;
                    break;
                }
                ind++;
            }
            if (switchToFormatString)
            {
                mode(CSharpLexer.INTERPOLATION_FORMAT);
            }
        }
    }

    void A6()
    {
        curlyLevels.add(1);
    }

    void A7()
    {
        interpolatedStringLevel--;
        interpolatedVerbatiums.removeLast();
        verbatium = (interpolatedVerbatiums.length > 0 ? interpolatedVerbatiums.last : false);
    }

    void A8()
    {
        curlyLevels.removeLast();
    }

    bool P1()
    {
        return !verbatium;
    }

    bool P2()
    {
        return verbatium;
    }
}

File: Go/csharp_lexer_base.go
package parser

import (
	"github.com/antlr/antlr4/runtime/Go/antlr"
)

// A simple integer stack

type IntStack []int

var ErrEmptyStack = "Stack is empty"

func (s *IntStack) Pop() int {
	l := len(*s) - 1
	if l < 0 {
		return 0
	}
	v := (*s)[l]
	*s = (*s)[0:l]
	return v
}

func (s *IntStack) Peek() int {
	l := len(*s) - 1
	if l < 0 {
		return 0
	}
	v := (*s)[l]
	return v
}

func (s *IntStack) Push(e int) {
	*s = append(*s, e)
}

func (s *IntStack) Count() int {
	return len(*s)
}

type BoolStack []bool

func (s *BoolStack) Pop() bool {
	l := len(*s) - 1
	if l < 0 {
		return false
	}
	v := (*s)[l]
	*s = (*s)[0:l]
	return v
}

func (s *BoolStack) Peek() bool {
	l := len(*s) - 1
	if l < 0 {
		return false
	}
	v := (*s)[l]
	return v
}

func (s *BoolStack) Push(e bool) {
	*s = append(*s, e)
}

func (s *BoolStack) Count() int {
	return len(*s)
}

// CSharpLexerBase implementation.
type CSharpLexerBase struct {
	*antlr.BaseLexer
	interpolatedStringLevel int
	interpolatedVerbatiums  BoolStack
	curlyLevels             IntStack
	verbatium               bool
}

func (p *CSharpLexer) A1() {
	x := p.CSharpLexerBase
	x.interpolatedStringLevel++
	x.interpolatedVerbatiums.Push(false)
	x.verbatium = false
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) A2() {
	x := p.CSharpLexerBase
	x.interpolatedStringLevel++
	x.interpolatedVerbatiums.Push(true)
	x.verbatium = true
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) A3() {
	x := p.CSharpLexerBase
	if x.interpolatedStringLevel > 0 {
		x.curlyLevels.Push(x.curlyLevels.Pop() + 1)
	}
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) A4() {
	x := p.CSharpLexerBase
	if x.interpolatedStringLevel > 0 {
		x.curlyLevels.Push(x.curlyLevels.Pop() - 1)
		if x.curlyLevels.Peek() == 0 {
			x.curlyLevels.Pop()
			p.Skip()
			p.PopMode()
		}
	}
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) A5() {
	x := p.CSharpLexerBase
	if x.interpolatedStringLevel > 0 {
		ind := 1
		switchToFormatString := true
		for x.GetInputStream().LA(ind) != '}' {
			if x.GetInputStream().LA(ind) == ':' || x.GetInputStream().LA(ind) == ')' {
				switchToFormatString = false
				break
			}
			ind++
		}
		if switchToFormatString {
			p.SetMode(CSharpLexerINTERPOLATION_FORMAT)
		}
	}
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) A6() {
	x := p.CSharpLexerBase
	x.curlyLevels.Push(1)
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) A7() {
	x := p.CSharpLexerBase
	x.interpolatedStringLevel--
	x.interpolatedVerbatiums.Pop()
	if x.interpolatedVerbatiums.Count() > 0 {
		x.verbatium = x.interpolatedVerbatiums.Peek()
	} else {
		x.verbatium = false
	}
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) A8() {
	x := p.CSharpLexerBase
	x.curlyLevels.Pop()
	p.CSharpLexerBase = x
}

func (p *CSharpLexer) P1() bool {
	x := p.CSharpLexerBase
	return !x.verbatium
}

func (p *CSharpLexer) P2() bool {
	x := p.CSharpLexerBase
	return x.verbatium
}
